### PR TITLE
Chore/anton refactor classifying code

### PR DESCRIPTION
**Pull request by antngh**
Refactoring code to make it much simpler to run the classification over models/layers/datasets.

Still a WIP, but should be much more streamlined to gather lots of stats. We can add in visualisation code separately in notebooks to generate plots for the parts we care about. This should also make it much easier to create the plots that we want.

### CODE REVIEW

File: src/stages/visualise_classification.py [hunk 1]
```diff
@@ -11,105 +11,156 @@
 import torch.nn as nn
```
Comment (antngh) [inline src/stages/visualise_classification.py:11]: I'll rename this, this stage will be about running the algorithms and storing stats. Most of the commented out code will go. The plotting code will go somewhere else to make it nicely callable from another stage or notebooj.

File: src/stages/visualise_classification.py [hunk 2]
```diff
@@ -386,99 +464,173 @@ def get_experiment_path(output_path, _experiment_id=None):
...
+    X_test = activations[test_indices]
+    return (X_train, train_labels), (X_test, test_labels)
+
+
+def llama3_8b_cities_augmented_data_loader():
+    pass
+
+
+# todo put this in a config.
+model_data_loaders = {
```
Comment (antngh) [inline src/stages/visualise_classification.py:506]: I'm thinking if we have some different functions for loading the activations and labels depending on the model and dataset (can have the same function over multiple if there is nothing complex). For example for the cities we have some special logic to remove IDK

File: src/stages/visualise_classification.py [hunk 3]
```diff
@@ -386,99 +464,173 @@ def get_experiment_path(output_path, _experiment_id=None):
...
+    }
+}
+
+
 def classifier_experiment_run(
-    model: str,
-    layer: str,
-    activations_dir: str,
-    question_answer_file: str,
+    run_config: list(tuple(str, str, str)),
```
Comment (antngh) [inline src/stages/visualise_classification.py:514]: would like this to run over models, datasets, layers, whatever you define. The idea being that it would run over a whole range. For example for a given model and layer you might want to run over a bunch of datasets, or for a given model and dataset you might want to run over all layers (or just over everything.

File: src/stages/visualise_classification.py [hunk 4]
```diff
@@ -386,99 +464,173 @@ def get_experiment_path(output_path, _experiment_id=None):
...
+                    X_test=X_test,
+                    train_labels=train_labels.to_numpy(),
+                    test_labels=test_labels.to_numpy(),
+                ).classification_metrics
+            )
+
+        # with (experiment_path / "info.json").open("w", encoding="utf-8") as file:
+        #     json.dump(stats_dict, file, indent=4, ensure_ascii=False)
+        overall_stats_dict[(model_name, dataset_name, f"layer_{layer}")] = stats_dict
+    return stats_dict
```
Comment (antngh) [inline src/stages/visualise_classification.py:632]: The output is just a summary dict that you can then save. We can then dig in and generate plots just for the interesting ones in the next stage or in a notebook,

File: src/classifying/activations.py [hunk 5]
```diff
@@ -0,0 +1,24 @@
+from typing import Any
```
Comment (antngh) [inline src/classifying/activations.py:1]: dont worry will think of better names and will properly document everything

File: src/classifying/activations.py [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+
+import torch as pt
+
+
+class Activations:
```
Comment (antngh) [inline src/classifying/activations.py:6]: util for getting the activations corresponding labels

File: src/classifying/classification_utils.py [hunk 7]
```diff
@@ -0,0 +1,130 @@
...
+            self.train_true_positive_rate,
+            self.train_thresholds,
+        ) = roc_curve(self.train_labels, self.train_classification_score)
+        self.test_false_positive_rate, self.test_true_positive_rate, _ = roc_curve(
+            self.test_labels, self.test_classification_score
+        )
+        self.roc_auc = auc(self.test_false_positive_rate, self.test_true_positive_rate)
+        self.classification_metric_funcs = classification_metric_funcs
+
+        self.classification_metrics = {
```
Comment (antngh) [inline src/classifying/classification_utils.py:117]: This is what we care about for the purposes of stats collecting. It's still useful to have this as a class though because later for visualising and other functionality, the other properties will be super useful.

**Review by ivanvmoreno — APPROVED** [review]
LGTM
